### DISCUSSION TITLE
Prefetching, side-effects and refreshing other routes

### DISCUSSION THREAD

**Discussion by johtso** [erikras/remix-conf-2022#4] [General]
Firstly, immensely enjoyed your talks, the whole concept here is super exciting, especially for those that have drunk the state machine cool aid!

## Cookies, prefetching and side-effects in loaders

One issue I see with using a cookie to store the state is that the `loader` now has side effects. Just hitting a url causes the state machine to be interpreted and the state to change (because our state machine controls our url).

This means we can't have prefetching, one of the awesome things Remix gives us. As soon as we prefetch the page the cookie will be updated with the new state, even if we never end up performing that navigation.

I wonder if an idea would be to use localstorage instead of the cookie, and update the localstorage state when the route is hydrated, and send it with each request. I guess that again breaks URL/history based navigation..

## Nested routes

Another snag I bumped into was using another awesome Remix feature, nested routes.

Because the state is being updated by a navigation rather than an action, Remix doesn't know to refresh other routes on the page that also use the state machine.. for example a base route that displays something different depending on what page we're on, like a navigation tab bar that shows the current page in a different colour.

But if we do all navigation using actions / form submissions, then we don't get browser history navigation.

There's probably some workaround here to force refresh all routes on the page, but I think it involves javascript triggering a reload after the route has hydrated.

## Prefetching _any_ event?

I'm definitely also imagining a scenario where in our state machine we declare which states / events are "safe". We can then prefetch any possible action, and only use the resulting new state if the user ends up actually performing that action. I'm not sure how you would do this as Remix doesn't have a concept of prefetching form submissions though..

## Misc..

One thing that I was super stoked by when playing around with this pattern was that we could get really nice type safety! And also the other nice things that xstate gives you like knowing if a particular event is possible given the current state (should a button be disabled).

```typescript
interface EventButtonProps {
  machine: InterpreterFrom<typeof bookingMachine>;
  event: XStateEvent<BookingMachineEvent>;
}

export function EventButton(
    { machine, event, className, ...props}:
    EventButtonProps &
    React.HTMLAttributes<HTMLButtonElement>
  ) {
  event = toEventObject(event);
  const isPossible = machine.state.can(event);

  return (
    <Form method="post">
      <button
        className={clsx(
          className,
          "event-button",
        )}
        type="submit"
        name="event.type"
        value={event.type}
        data-disabled={!isPossible}
        {...props}
      />
    </Form>
  )
};
```

**Comment by erikras** [comment]
These are all valid points worthy of discussion. This repo is just a proof of concept for now.
